### PR TITLE
Factor out NextScript inline source (#4934)

### LINKED ISSUE #4934: Factor out `NextScript` inline source to be able to generate hash to employ a CSP
# Feature request

## Is your feature request related to a problem? Please describe.
I want to employ a proper CSP for pages that are statically exported and served by a CDN. The `script-src` currently has to allow `unsafe-inline` because of inline scripts being defined per `NextScript`.

## Describe the solution you'd like
I want to factor out the inline script source generation of `NextScript` to be accessible as a static method. Then it can e.g. be used in a custom `_document` to generate the hash and emit the CSP `meta` tag with it.

## Describe alternatives you've considered
An alternative would be to actually not inline the `NextScript` script, but load it as a file.
As stated, since nonces need to be generated / not guessable, this is not an option for CDN deployments.

## Additional context
https://developers.google.com/web/fundamentals/security/csp/

### PR DESCRIPTION
**Pull request by dbo**
This PR factors out the inline script into an own static method; fixes #4934.

### CODE REVIEW

File: server/document.js [hunk 1]
```diff
@@ -162,9 +162,42 @@ export class NextScript extends Component {
...
+      __NEXT_DATA__ = ${htmlescape(__NEXT_DATA__)}
+      module={}
+      __NEXT_LOADED_PAGES__ = []
+      __NEXT_LOADED_CHUNKS__ = []
+
+      __NEXT_REGISTER_PAGE = function (route, fn) {
+        __NEXT_LOADED_PAGES__.push({ route: route, fn: fn })
+      }
+
+      __NEXT_REGISTER_CHUNK = function (chunkName, fn) {
```
Comment (timneutkens) [inline server/document.js]: For some reason, you're adding back obsolete code that was removed in latest canary, maybe you created a branch based on master.

Here's the correct code: https://github.com/zeit/next.js/blob/canary/server/document.js#L186-L200

**Comment by dav-is** [comment]
It doesn’t appear that this PR would stop the execution of an inline script. It simply moves the logic that generates the inline code. It still generates the <script> with the inline script that’s executed by the browser even if it’s not “inline” in the react sense.

File: server/document.js (hunk 1 above)
Comment (dbo) [reply server/document.js]: Yes, I was not aware that development is not happening on master, I will move to the canary branch.

**Comment by dbo** [comment]
@dav-is Yes, it's still inline, the alternative would be to load it as a generated file. But exposing the script source will allow to generate a CSP hash for it and emit this per custom `_document` `meta` tag.

**Pull request closed without merging**